File: app/src/main/java/life/plank/juna/zone/data/network/interfaces/RestApi.java
package life.plank.juna.zone.data.network.interfaces;

import com.google.gson.JsonObject;

import java.util.Date;
import java.util.List;
import java.util.Set;

import life.plank.juna.zone.data.model.Board;
import life.plank.juna.zone.data.model.Emoji;
import life.plank.juna.zone.data.model.FeedEntry;
import life.plank.juna.zone.data.model.FeedItem;
import life.plank.juna.zone.data.model.FeedItemComment;
import life.plank.juna.zone.data.model.FootballTeam;
import life.plank.juna.zone.data.model.Lineups;
import life.plank.juna.zone.data.model.MatchDetails;
import life.plank.juna.zone.data.model.MatchFixture;
import life.plank.juna.zone.data.model.MatchStats;
import life.plank.juna.zone.data.model.NextMatch;
import life.plank.juna.zone.data.model.PlayerStats;
import life.plank.juna.zone.data.model.ScrubberData;
import life.plank.juna.zone.data.model.SignUpModel;
import life.plank.juna.zone.data.model.Standings;
import life.plank.juna.zone.data.model.TeamStats;
import life.plank.juna.zone.data.model.User;
import life.plank.juna.zone.data.model.Zones;
import life.plank.juna.zone.data.model.card.JunaCard;
import life.plank.juna.zone.data.model.notification.CardNotification;
import life.plank.juna.zone.data.model.notification.SocialNotification;
import life.plank.juna.zone.data.model.poll.Poll;
import life.plank.juna.zone.data.model.poll.PollAnswerRequest;
import life.plank.juna.zone.data.model.poll.PollAnswerResponse;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Multipart;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * Created by plank-dhamini on 6/8/2017.
 * This class consists of all the rest api calls for zone
 */

public interface RestApi {

    String FOOTBALL_SUFFIX = "football";
    String ZONE_BACKEND_SUFFIX = "backend";
    String CARD_SUFFIX = "card";
    String FEED_FLOWS_SUFFIX = "feedflows";

    //working
    @GET(FOOTBALL_SUFFIX + "/seasons/standings")
    Observable<Response<List<Standings>>> getStandings(@Query("leagueName") String leagueName,
                                                       @Query("seasonName") String seasonName,
                                                       @Query("countryName") String countryName);

    //working
    @GET(FOOTBALL_SUFFIX + "/seasons/playerstats")
    Observable<Response<List<PlayerStats>>> getPlayerStats(@Query("leagueName") String leagueName,
                                                           @Query("seasonName") String seasonName,
                                                           @Query("countryName") String countryName);

    //working
    @GET(FOOTBALL_SUFFIX + "/teams/stats")
    Observable<Response<List<TeamStats>>> getTeamStats(@Query("leagueName") String leagueName,
                                                       @Query("seasonName") String seasonName,
                                                       @Query("countryName") String countryName);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards")
    Observable<Response<Board>> getBoard(@Query("foreignId") Long matchId, @Query("boardType") String boardType, @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards/createdBy")
    Observable<Response<List<Board>>> getUserBoards(@Header("Authorization") String authHeader);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/boards/{id}/invite")
    Observable<Response<JsonObject>> inviteUserToJoinBoard(@Body Set<User> user, @Path("id") String boardId, @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards/feedItems")
    Observable<Response<List<FeedEntry>>> getBoardFeedItems(@Query("id") String boardId, @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/zones")
    Observable<Response<List<Zones>>> retrieveZones();

    //working
    @DELETE(ZONE_BACKEND_SUFFIX + "/boards/{id}/activities/removeUser")
    Observable<Response<JsonObject>> deleteUserFromPrivateBoard(@Path("id") String boardId,
                                                                @Query("boardUserId") String userId,
                                                                @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/users")
    Observable<Response<User>> getUser(@Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/users/editUserDetails")
    Observable<Response<JsonObject>> updateUserDetails(@Header("Authorization") String authHeader,
                                                       @Body User user);

    //working
    @Multipart
    @POST(ZONE_BACKEND_SUFFIX + "/activities/upload")
    Observable<Response<JsonObject>> postMediaContentToServer(@Part MultipartBody.Part file,
                                                              @Query("targetId") String targetId,
                                                              @Query("contentType") String contentType,
                                                              @Query("userId") String userId,
                                                              @Query("dateCreated") String dateCreated,
                                                              @Query("feedType") String feedType,
                                                              @Query("title") String title,
                                                              @Header("Authorization") String authHeader);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/activities/feeditems")
    Observable<Response<JsonObject>> postFeedItemOnBoard(@Body FeedItem feedItem,
                                                         @Query("boardId") String boardId,
                                                         @Query("contentType") String contentType,
                                                         @Header("Authorization") String authHeader);

    //working
    @Multipart
    @POST(ZONE_BACKEND_SUFFIX + "/boards")
    Observable<Response<String>> createPrivateBoard(@Query("boardType") String boardType,
                                                    @Part("displayName") RequestBody displayName,
                                                    @Part("zone") RequestBody zone,
                                                    @Part("description") RequestBody description,
                                                    @Part("color") RequestBody color,
                                                    @Part MultipartBody.Part file,
                                                    @Header("Authorization") String authHeader);

    //working
    @Multipart
    @POST(ZONE_BACKEND_SUFFIX + "/users/profilePicture")
    Observable<Response<String>> uploadProfilePicture(@Part MultipartBody.Part file,
                                                      @Header("Authorization") String authHeader);

    //working
    @DELETE(ZONE_BACKEND_SUFFIX + "/boards/{id}")
    Observable<Response<JsonObject>> deleteBoard(@Path("id") String boardId,
                                                 @Header("Authorization") String authHeader);

    //working
    @GET(FOOTBALL_SUFFIX + "/seasons/matches")
    Observable<Response<List<MatchFixture>>> getFixtures(@Query("seasonName") String seasonName,
                                                         @Query("leagueName") String leagueName,
                                                         @Query("countryName") String countryName);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/feedEntries")
    Observable<Response<List<FeedEntry>>> getUserFeed(@Header("Authorization") String... authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards/{id}")
    Observable<Response<Board>> getBoardById(@Path("id") String boardId, @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/activities/{feedItemId}/emojiCounts")
    Observable<Response<List<Emoji>>> getTopFeedItemEmoji(@Path("feedItemId") String feedItemId, @Header("Authorization") String authHeader);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/activities/{feedItemId}/emojis")
    Observable<Response<JsonObject>> postEmojiOnFeedItem(@Path("feedItemId") String feedItemId,
                                                         @Query("boardId") String boardId,
                                                         @Query("reaction") Integer emojiUnicode,
                                                         @Query("time") String dateCreated,
                                                         @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/emojiCounts")
    Observable<Response<List<Emoji>>> getTopBoardEmoji(@Path("boardId") String boardId, @Header("Authorization") String authHeader);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/emojis")
    Observable<Response<JsonObject>> postEmojiOnBoard(@Path("boardId") String boardId,
                                                      @Query("reaction") Integer emojiUnicode,
                                                      @Query("time") String dateCreated,
                                                      @Header("Authorization") String authHeader);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/users/search")
    Observable<Response<List<User>>> getSearchedUsers(@Header("Authorization") String authHeader, @Query("q") String displayName);

    //working
    @GET(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/members")
    Observable<Response<List<User>>> getBoardMembers(@Path("boardId") String boardId, @Header("Authorization") String authHeader);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/activities/follow")
    Observable<Response<JsonObject>> followBoard(@Header("Authorization") String authHeader, @Path("boardId") String boardId);

    //working
    @POST(ZONE_BACKEND_SUFFIX + "/zones/follows")
    Observable<Response<JsonObject>> followZones(@Header("Authorization") String authHeader, @Body Zones zones);

    @GET(ZONE_BACKEND_SUFFIX + "/boards/following")
    Observable<Response<List<Board>>> getFollowingBoards(@Query("zone") String zone, @Header("Authorization") String authHeader);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}/lineups")
    Observable<Response<Lineups>> getLineUpsData(@Path("matchId") long matchId);

    @POST(ZONE_BACKEND_SUFFIX + "/ausers")
    Observable<Response<SignUpModel>> createUser(@Body SignUpModel signUpModel);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}/stats")
    Observable<Response<MatchStats>> getMatchStatsForMatch(@Path("matchId") long matchId);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}")
    Observable<Response<MatchDetails>> getMatchDetails(@Path("matchId") long matchId);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}/standings")
    Observable<Response<List<Standings>>> getMatchStandingsForMatch(@Path("matchId") long matchId);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}/teamStats")
    Observable<Response<List<TeamStats>>> getTeamStatsForMatch(@Path("matchId") long matchId);

    @GET(FOOTBALL_SUFFIX + "/matches/{matchId}/scrubber/{hour}")
    Observable<Response<List<ScrubberData>>> getScrubberDetails(@Path("matchId") long matchId, @Path("hour") Date currentMatchTime);

    @POST(ZONE_BACKEND_SUFFIX + "/activities/{id}/pins")
    Observable<Response<String>> pinFeedItem(@Path("id") String feedItemId,
                                             @Query("target") String target,
                                             @Query("targetId") String boardId,
                                             @Query("time") String dateCreated,
                                             @Header("Authorization") String authHeader);

    @DELETE(ZONE_BACKEND_SUFFIX + "/activities/{id}/pins/{pinId}")
    Observable<Response<JsonObject>> unpinFeedItem(@Path("id") String boardId, @Path("pinId") String pinId, @Header("Authorization") String authHeader);

    @GET(ZONE_BACKEND_SUFFIX + "/activities/{feedItemId}/comments")
    Observable<Response<List<FeedItemComment>>> getCommentsForFeed(@Path("feedItemId") String feedId,
                                                                   @Header("Authorization") String authHeader);

    @GET(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/comments")
    Observable<Response<List<FeedItemComment>>> getCommentsForBoard(@Path("boardId") String boardId,
                                                                    @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/comments")
    Observable<Response<FeedItemComment>> postCommentOnBoard(@Body String comment,
                                                             @Path("boardId") String boardId,
                                                             @Query("time") String time,
                                                             @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/comments/{commentId}")
    Observable<Response<FeedItemComment>> postReplyOnBoardComment(@Body String reply,
                                                                  @Path("commentId") String commentId,
                                                                  @Path("boardId") String boardId,
                                                                  @Query("time") String time,
                                                                  @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/activities/{feedItemId}/comments")
    Observable<Response<FeedItemComment>> postCommentOnFeedItem(@Body String comment,
                                                                @Path("feedItemId") String feedItemId,
                                                                @Query("boardId") String boardId,
                                                                @Query("time") String time,
                                                                @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/activities/{feedItemId}/comments/{commentId}")
    Observable<Response<FeedItemComment>> postReplyOnComment(@Body String reply,
                                                             @Path("feedItemId") String feedItemId,
                                                             @Path("commentId") String commentId,
                                                             @Query("boardId") String boardId,
                                                             @Query("time") String time,
                                                             @Header("Authorization") String authHeader);

    @GET(FEED_FLOWS_SUFFIX + "/polls/getPoll")
    Observable<Response<Poll>> getBoardPoll(@Query("boardId") String boardId, @Header("Authorization") String authHeader);

    @GET(FEED_FLOWS_SUFFIX + "/polls/userSelection")
    Observable<Response<PollAnswerResponse>> getBoardPollAnswer(@Query("pollId") long pollId, @Header("Authorization") String authHeader);

    @POST(FEED_FLOWS_SUFFIX + "/polls/userSelection")
    Observable<Response<PollAnswerResponse>> postBoardPollAnswer(@Body PollAnswerRequest pollAnswer, @Header("Authorization") String authHeader);

    @GET(FOOTBALL_SUFFIX + "/teams")
    Observable<Response<List<FootballTeam>>> getSearchedFootballTeams(@Query("teamNamePart") String teamName, @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/clap")
    Observable<Response<JsonObject>> postClap(@Path("boardId") String boardId, @Body User player);

    @POST(ZONE_BACKEND_SUFFIX + "/boards/{boardId}/dart")
    Observable<Response<JsonObject>> postDart(@Path("boardId") String boardId, @Body User player);

    @POST(ZONE_BACKEND_SUFFIX + "/users/zonePreferences")
    Observable<Response<String>> postTeamPreferences(@Query("zone") String zone,
                                                     @Body Set<String> teamList,
                                                     @Header("Authorization") String authHeader);

    @GET(FOOTBALL_SUFFIX + "/teams/popular")
    Observable<Response<List<FootballTeam>>> getPopularTeams(@Header("Authorization") String authHeader);

    @GET(ZONE_BACKEND_SUFFIX + "/feedEntries/{id}")
    Observable<Response<FeedEntry>> getFeedEntry(@Path("id") String feedItemId, @Header("Authorization") String authHeader);

    @GET(ZONE_BACKEND_SUFFIX + "/notifications/unread")
    Observable<Response<List<SocialNotification>>> getNotifications(@Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/notifications/{id}/setIsRead")
    Observable<Response<List<SocialNotification>>> setNotificationAsRead(@Path("id") String notificationId, @Header("Authorization") String authHeader);

    @POST(ZONE_BACKEND_SUFFIX + "/notifications/{id}/setIsReadAfter")
    Observable<Response<List<SocialNotification>>> setAllNotificationsAsRead(@Path("id") String firstNotificationId, @Header("Authorization") String authHeader);

    @GET(FOOTBALL_SUFFIX + "/matches/nextMatches")
    Observable<Response<List<NextMatch>>> getNextMatches(@Query("leagues") List<String> userSelectedLeagues, @Header("Authorization") String authHeader);

    @GET(CARD_SUFFIX + "/cards/wallet")
    Observable<Response<List<JunaCard>>> getCardWallet(@Header("Authorization") String authHeader);

    @GET(CARD_SUFFIX + "/cards/{id}")
    Observable<Response<JunaCard>> getCardDetail(@Path("id") String cardId, @Header("Authorization") String authHeader);

    @Multipart
    @POST(CARD_SUFFIX + "/cards")
    Observable<Response<JunaCard>> createCard(@Query("cardColor") String cardColor,
                                              @Part MultipartBody.Part image,
                                              @Header("Authorization") String authHeader);

    @Multipart
    @PATCH(CARD_SUFFIX + "/cards")
    Observable<Response<JunaCard>> updateCard(@Part("cardColor") String cardColor,
                                              @Part MultipartBody.Part image,
                                              @Header("Authorization") String authHeader);

    @PUT(CARD_SUFFIX + "/cards/{id}/publish")
    Observable<Response<Void>> publishCard(@Path("id") String cardId, @Header("Authorization") String authHeader);

    @GET(CARD_SUFFIX + "/cardNotifications")
    Observable<Response<List<CardNotification>>> getCardNotifications(@Header("Authorization") String authHeader);

    @PATCH(CARD_SUFFIX + "/cardNotifications")
    Observable<Response<Void>> markCardNotificationRead(@Header("Authorization") String authHeader);
}

File: app/src/main/java/life/plank/juna/zone/data/api/RestApi.kt
package life.plank.juna.zone.data.api

import com.google.gson.JsonObject
import life.plank.juna.zone.data.model.board.*
import life.plank.juna.zone.data.model.board.poll.*
import life.plank.juna.zone.data.model.card.*
import life.plank.juna.zone.data.model.feed.*
import life.plank.juna.zone.data.model.football.*
import life.plank.juna.zone.data.model.notification.*
import life.plank.juna.zone.data.model.user.*
import life.plank.juna.zone.sharedpreference.IdToken
import okhttp3.*
import retrofit2.Response
import retrofit2.http.*
import rx.Observable
import java.util.*

const val FOOTBALL_SUFFIX = "football"
const val ZONE_BACKEND_SUFFIX = "backend"
const val CARD_SUFFIX = "card"
const val FEED_FLOWS_SUFFIX = "feedflows"

/**
 * This class consists of all the rest api calls for zone
 */
interface RestApi {

    //working
    @GET("$FOOTBALL_SUFFIX/seasons/standings")
    fun getStandings(@Query("leagueName") leagueName: String,
                     @Query("seasonName") seasonName: String,
                     @Query("countryName") countryName: String): Observable<Response<MutableList<Standings>>>

    //working
    @GET("$FOOTBALL_SUFFIX/seasons/playerstats")
    fun getPlayerStats(@Query("leagueName") leagueName: String,
                       @Query("seasonName") seasonName: String,
                       @Query("countryName") countryName: String): Observable<Response<MutableList<PlayerStats>>>

    //working
    @GET("$FOOTBALL_SUFFIX/teams/stats")
    fun getTeamStats(@Query("leagueName") leagueName: String,
                     @Query("seasonName") seasonName: String,
                     @Query("countryName") countryName: String): Observable<Response<MutableList<TeamStats>>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards")
    fun getBoard(@Query("foreignId") matchId: Long?, @Query("boardType") boardType: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<Board>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards/createdBy")
    fun getUserBoards(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<Board>>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/boards/{id}/invite")
    fun inviteUserToJoinBoard(@Body user: Set<User>, @Path("id") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards/feedItems")
    fun getBoardFeedItems(@Query("id") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FeedEntry>>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/zones")
    fun retrieveZones(): Observable<Response<MutableList<Zones>>>

    //working
    @DELETE("$ZONE_BACKEND_SUFFIX/boards/{id}/activities/removeUser")
    fun deleteUserFromPrivateBoard(@Path("id") boardId: String,
                                   @Query("boardUserId") userId: String,
                                   @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/users")
    fun getUser(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<User>>

    @POST("$ZONE_BACKEND_SUFFIX/users/editUserDetails")
    fun updateUserDetails(@Body user: User, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @Multipart
    @POST("$ZONE_BACKEND_SUFFIX/activities/upload")
    fun postMediaContentToServer(@Part file: MultipartBody.Part,
                                 @Query("targetId") targetId: String?,
                                 @Query("contentType") contentType: String,
                                 @Query("userId") userId: String,
                                 @Query("dateCreated") dateCreated: String,
                                 @Query("feedType") feedType: String,
                                 @Query("title") title: String,
                                 @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/activities/feeditems")
    fun postFeedItemOnBoard(@Body feedItem: FeedItem,
                            @Query("boardId") boardId: String?,
                            @Query("contentType") contentType: String,
                            @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @Multipart
    @POST("$ZONE_BACKEND_SUFFIX/boards")
    fun createPrivateBoard(@Query("boardType") boardType: String,
                           @Part("displayName") displayName: RequestBody,
                           @Part("zone") zone: RequestBody,
                           @Part("description") description: RequestBody,
                           @Part("color") color: RequestBody,
                           @Part file: MultipartBody.Part,
                           @Header("Authorization") authHeader: String? = IdToken): Observable<Response<String>>

    //working
    @Multipart
    @POST("$ZONE_BACKEND_SUFFIX/users/profilePicture")
    fun uploadProfilePicture(@Part file: MultipartBody.Part, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<String>>

    //working
    @DELETE("$ZONE_BACKEND_SUFFIX/boards/{id}")
    fun deleteBoard(@Path("id") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @GET("$FOOTBALL_SUFFIX/seasons/matches")
    fun getFixtures(@Query("seasonName") seasonName: String,
                    @Query("leagueName") leagueName: String,
                    @Query("countryName") countryName: String): Observable<Response<MutableList<MatchFixture>>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/feedEntries")
    fun getUserFeed(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FeedEntry>>>

    @GET("$ZONE_BACKEND_SUFFIX/feedEntries")
    fun getAnonymousFeed(): Observable<Response<MutableList<FeedEntry>>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards/{id}")
    fun getBoardById(@Path("id") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<Board>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/activities/{feedItemId}/emojiCounts")
    fun getTopFeedItemEmoji(@Path("feedItemId") feedItemId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<Emoji>>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/activities/{feedItemId}/emojis")
    fun postEmojiOnFeedItem(@Path("feedItemId") feedItemId: String,
                            @Query("boardId") boardId: String,
                            @Query("reaction") emojiUnicode: Int?,
                            @Query("time") dateCreated: String,
                            @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards/{boardId}/emojiCounts")
    fun getTopBoardEmoji(@Path("boardId") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<Emoji>>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/emojis")
    fun postEmojiOnBoard(@Path("boardId") boardId: String,
                         @Query("reaction") emojiUnicode: Int?,
                         @Query("time") dateCreated: String,
                         @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/users/search")
    fun getSearchedUsers(@Query("q") displayName: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<User>>>

    //working
    @GET("$ZONE_BACKEND_SUFFIX/boards/{boardId}/members")
    fun getBoardMembers(@Path("boardId") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<User>>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/activities/follow")
    fun followBoard(@Path("boardId") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    //working
    @POST("$ZONE_BACKEND_SUFFIX/zones/follows")
    fun followZones(@Body zones: Zones, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    @GET("$ZONE_BACKEND_SUFFIX/boards/following")
    fun getFollowingBoards(@Query("zone") zone: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<Board>>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}/lineups")
    fun getLineUpsData(@Path("matchId") matchId: Long): Observable<Response<Lineups>>

    @POST("$ZONE_BACKEND_SUFFIX/ausers")
    fun createUser(@Body signUpModel: SignUpModel): Observable<Response<SignUpModel>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}/stats")
    fun getMatchStatsForMatch(@Path("matchId") matchId: Long): Observable<Response<MatchStats>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}")
    fun getMatchDetails(@Path("matchId") matchId: Long): Observable<Response<MatchDetails>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}/standings")
    fun getMatchStandingsForMatch(@Path("matchId") matchId: Long): Observable<Response<MutableList<Standings>>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}/teamStats")
    fun getTeamStatsForMatch(@Path("matchId") matchId: Long): Observable<Response<MutableList<TeamStats>>>

    @GET("$FOOTBALL_SUFFIX/matches/{matchId}/scrubber/{hour}")
    fun getScrubberDetails(@Path("matchId") matchId: Long, @Path("hour") currentMatchTime: Date): Observable<Response<MutableList<ScrubberData>>>

    @POST("$ZONE_BACKEND_SUFFIX/activities/{id}/pins")
    fun pinFeedItem(@Path("id") feedItemId: String,
                    @Query("target") target: String,
                    @Query("targetId") boardId: String,
                    @Query("time") dateCreated: String,
                    @Header("Authorization") authHeader: String? = IdToken): Observable<Response<String>>

    @DELETE("$ZONE_BACKEND_SUFFIX/activities/{id}/pins/{pinId}")
    fun unpinFeedItem(@Path("id") boardId: String, @Path("pinId") pinId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JsonObject>>

    @GET("$ZONE_BACKEND_SUFFIX/activities/{feedItemId}/comments")
    fun getCommentsForFeed(@Path("feedItemId") feedId: String,
                           @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FeedItemComment>>>

    @GET("$ZONE_BACKEND_SUFFIX/boards/{boardId}/comments")
    fun getCommentsForBoard(@Path("boardId") boardId: String,
                            @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FeedItemComment>>>

    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/comments")
    fun postCommentOnBoard(@Body comment: String,
                           @Path("boardId") boardId: String,
                           @Query("time") time: String,
                           @Header("Authorization") authHeader: String? = IdToken): Observable<Response<FeedItemComment>>

    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/comments/{commentId}")
    fun postReplyOnBoardComment(@Body reply: String,
                                @Path("commentId") commentId: String,
                                @Path("boardId") boardId: String,
                                @Query("time") time: String,
                                @Header("Authorization") authHeader: String? = IdToken): Observable<Response<FeedItemComment>>

    @POST("$ZONE_BACKEND_SUFFIX/activities/{feedItemId}/comments")
    fun postCommentOnFeedItem(@Body comment: String,
                              @Path("feedItemId") feedItemId: String,
                              @Query("boardId") boardId: String,
                              @Query("time") time: String,
                              @Header("Authorization") authHeader: String? = IdToken): Observable<Response<FeedItemComment>>

    @POST("$ZONE_BACKEND_SUFFIX/activities/{feedItemId}/comments/{commentId}")
    fun postReplyOnComment(@Body reply: String,
                           @Path("feedItemId") feedItemId: String,
                           @Path("commentId") commentId: String,
                           @Query("boardId") boardId: String,
                           @Query("time") time: String,
                           @Header("Authorization") authHeader: String? = IdToken): Observable<Response<FeedItemComment>>

    @GET("$FEED_FLOWS_SUFFIX/polls/getPoll")
    fun getBoardPoll(@Query("boardId") boardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<Poll>>

    @GET("$FEED_FLOWS_SUFFIX/polls/userSelection")
    fun getBoardPollAnswer(@Query("pollId") pollId: Long, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<PollAnswerResponse>>

    @POST("$FEED_FLOWS_SUFFIX/polls/userSelection")
    fun postBoardPollAnswer(@Body pollAnswer: PollAnswerRequest, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<PollAnswerResponse>>

    @GET("$FOOTBALL_SUFFIX/teams")
    fun getSearchedFootballTeams(@Query("teamNamePart") teamName: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FootballTeam>>>

    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/clap")
    fun postClap(@Path("boardId") boardId: String, @Body player: User): Observable<Response<JsonObject>>

    @POST("$ZONE_BACKEND_SUFFIX/boards/{boardId}/dart")
    fun postDart(@Path("boardId") boardId: String, @Body player: User): Observable<Response<JsonObject>>

    @POST("$ZONE_BACKEND_SUFFIX/users/zonePreferences")
    fun postTeamPreferences(@Query("zone") zone: String,
                            @Body teamList: Set<String>,
                            @Header("Authorization") authHeader: String? = IdToken): Observable<Response<String>>

    @GET("$FOOTBALL_SUFFIX/teams/popular")
    fun getPopularTeams(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<FootballTeam>>>

    @GET("$ZONE_BACKEND_SUFFIX/feedEntries/{id}")
    fun getFeedEntry(@Path("id") feedItemId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<FeedEntry>>

    @GET("$ZONE_BACKEND_SUFFIX/notifications/unread")
    fun getNotifications(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<SocialNotification>>>

    @POST("$ZONE_BACKEND_SUFFIX/notifications/{id}/setIsRead")
    fun setNotificationAsRead(@Path("id") notificationId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<SocialNotification>>>

    @POST("$ZONE_BACKEND_SUFFIX/notifications/{id}/setIsReadAfter")
    fun setAllNotificationsAsRead(@Path("id") firstNotificationId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<SocialNotification>>>

    @GET("$FOOTBALL_SUFFIX/matches/nextMatches")
    fun getNextMatches(@Query("leagues") userSelectedLeagues: List<String>, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<NextMatch>>>

    @GET("$CARD_SUFFIX/cards/wallet")
    fun getCardWallet(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<JunaCard>>>

    @GET("$CARD_SUFFIX/cards/{id}")
    fun getCardDetail(@Path("id") cardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JunaCard>>

    @Multipart
    @POST("$CARD_SUFFIX/cards")
    fun createCard(@Query("cardTemplate") cardTemplate: String,
                   @Part image: MultipartBody.Part,
                   @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JunaCardTemplate>>

    @Multipart
    @PATCH("$CARD_SUFFIX/cards")
    fun updateCard(@Part("cardTemplate") cardTemplate: String,
                   @Part image: MultipartBody.Part,
                   @Header("Authorization") authHeader: String? = IdToken): Observable<Response<JunaCardTemplate>>

    @PUT("$CARD_SUFFIX/cards/{id}/publish")
    fun publishCard(@Path("id") cardId: String, @Header("Authorization") authHeader: String? = IdToken): Observable<Response<Void>>

    @GET("$CARD_SUFFIX/cardNotifications")
    fun getCardNotifications(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<MutableList<CardNotification>>>

    @PATCH("$CARD_SUFFIX/cardNotifications")
    fun markCardNotificationRead(@Header("Authorization") authHeader: String? = IdToken): Observable<Response<Void>>
}